metals/src/main/scala/scala/meta/internal/metals/ScalafixProvider.scala

```diff
@@ -0,0 +1,184 @@
+import scala.meta.io.AbsolutePath
+
+import org.eclipse.lsp4j.MessageType
+import org.eclipse.{lsp4j => l}
+import scalafix.interfaces.Scalafix
+import scalafix.interfaces.ScalafixArguments
+
+case class ScalafixProvider(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+    scalafixConfigPath: RelativePath,
+    workspace: AbsolutePath,
+    embedded: Embedded,
+    statusBar: StatusBar,
+    icons: Icons,
+    languageClient: MetalsLanguageClient
+) {
+  import ScalafixProvider._
+  private val scalafixCache = TrieMap.empty[ScalaBinaryVersion, Scalafix]
+  private val organizeImportRuleCache =
+    TrieMap.empty[ScalaBinaryVersion, List[Path]]
+
+  def organizeImports(file: AbsolutePath): List[l.TextEdit] = {
+    val fileInput = file.toInput
+    val unsavedFile = file.toInputFromBuffers(buffers)
+    if (isUnsaved(unsavedFile.text, fileInput.text)) {
+      scribe.info(s"Organize imports requires saving the file first")
+      languageClient.showMessage(
+        MessageType.Warning,
+        s"Save ${file.toNIO.getFileName} before organizing imports"
+      )
+      Nil
+    } else {
+      if (Seq("scala", "sbt").contains(file.`extension`)) {
+        val scalafixConfPath = workspace.resolve(scalafixConfigPath)
+        val scalafixConf =
+          if (scalafixConfPath.isFile) Some(scalafixConfPath.toNIO)
+          else None
+        val scalafixEvaluation = for {
+          (scalaVersion, scalaBinaryVersion, classPath) <-
+            scalaVersionAndClasspath(file)
+          _ <-
+            if (!ScalaVersions.isScala3Version(scalaVersion)) Some(())
+            else {
+              scribe.warn(
+                s"Organize import doesn't work on $scalaVersion files"
+              )
+              None
+            }
+          api <- getOrUpdateScalafixCache(scalaBinaryVersion)
+          scalafixArgs = configureApi(api, scalaVersion, classPath)
+          urlClassLoaderWithExternalRule <- getOrUpdateRuleCache(
+            scalaBinaryVersion,
+            api.getClass.getClassLoader
+          )
+        } yield {
+          val scalacOption =
+            if (scalaBinaryVersion == "2.13") "-Wunused:imports"
+            else "-Ywarn-unused-import"
+
+          scalafixArgs
+            .withToolClasspath(urlClassLoaderWithExternalRule)
+            .withConfig(scalafixConf.asJava)
+            .withRules(List(organizeImportRuleName).asJava)
+            .withPaths(List(file.toNIO).asJava)
+            .withSourceroot(workspace.toNIO)
+            .withScalacOptions(Collections.singletonList(scalacOption))
+            .evaluate()
+        }
+
+        val newFileContentOpt = scalafixEvaluation
+          .flatMap(result => result.getFileEvaluations.headOption)
+          .flatMap(_.previewPatches().asScala)
+        newFileContentOpt.map(textEditsFrom(_, fileInput)).getOrElse(Nil)
+      } else {
+        scribe.info(
+          s"""|Could not organize import for ${file.toNIO.getFileName}. Should end with .scala or .sbt"""
```

Comment:
This should not be needed if we check the file extension while contributing the code action

Reply:
ok.. But It should be noted somewhere that we expect files to be .scala or .sbt ... as a comment ?

Reply:
We can leave it as a comment in the code action, but we also document it in the docs, so I think it should be fine